oximeter/db/src/oxql/plan/plan.rs

```diff
@@ -0,0 +1,1366 @@
+        grouped: &GroupedTableOp,
+        nodes: &mut Vec<Node>,
+    ) -> anyhow::Result<()> {
+        let mut subplans = Vec::with_capacity(grouped.ops.len());
+        for query in grouped.ops.iter().cloned() {
+            let start = Instant::now();
+            let mut new_nodes = Vec::new();
+            Self::plan_query(&query, schema, &mut new_nodes)?;
+            subplans.push(Plan {
+                query,
+                nodes: new_nodes.clone(),
+                optimized: OptimizedPlan::Unchanged(new_nodes),
+                duration: start.elapsed(),
+            });
+        }
+        nodes.push(Node::Subquery(subplans));
+        Ok(())
+    }
+
+    // Optimize a plan, if possible.
+    //
+    // This attempts to apply a number of plan-rewriting steps, to make the
+    // query as efficient as possible. Today these steps include:
+    //
+    // - Predicate pushdown: moving filtering predicates as close to the data as
+    // possible
+    // - Limit pushdown: Moving `first` or `last` table operations as close to
+    // the data as possible.
+    //
+    // There is a lot of room for new steps here. The most obvious next
+    // candidates are:
+    //
+    // - Deltas: push the computation of deltas from cumulative timeseries into
+    // the database
+    // - Alignment: push alignment operations into the database
+    // - Constant evaluation: As we support richer expressions, such as writing
+    // filters like `filter x > 1 / 2` or `filter x > log(5)`, we can evaluate
+    // those expressions at query-plan time.
+    fn optimize_plan(nodes: &[Node]) -> anyhow::Result<OptimizedPlan> {
+        let optimized = Self::pushdown_predicates(nodes)?;
+        Self::pushdown_limit(optimized.nodes())
+    }
+
+    // Push down limit operations in the list of plan nodes.
+    fn pushdown_limit(nodes: &[Node]) -> anyhow::Result<OptimizedPlan> {
+        anyhow::ensure!(
+            !nodes.is_empty(),
+            "Planning error: plan nodes cannot be empty"
+        );
+        let mut modified = false;
+
+        // Collect nodes in the plan.
+        let mut remaining_nodes =
+            nodes.iter().cloned().collect::<VecDeque<_>>();
+        let mut processed_nodes = VecDeque::with_capacity(nodes.len());
+
+        while let Some(current_node) = remaining_nodes.pop_back() {
+            // What we do with the limit node depends on what's in front of it.
+            let Some(next_node) = remaining_nodes.pop_back() else {
+                // If there _isn't_ one, then the current node must be the start
+                // of the plan, and so push it and break out.
+                processed_nodes.push_front(current_node);
+                break;
+            };
+
+            // If this isn't a limit node, just push it and continue
+            let Node::Limit(limit) = current_node else {
+                processed_nodes.push_front(current_node);
+                processed_nodes.push_front(next_node);
```

Comment:
```suggestion
                if matches!(next_node, Node::Limit(_)) {
                    remaining_nodes.push_back(next_node);
                } else {
                    processed_nodes.push_front(next_node);
                }
```

This skips `next_node`. If that is not itself a limit, we _can_ skip it in this method. But if it _is_ a limit, we need to process it on the next pass through the loop, so push it onto the list of remaining nodes.

Alternatively, if this seems like premature optimization, we can _always_ push this onto the remaining nodes. It will be processed the next time through the loop, and handled in that case.